### PR TITLE
 ENT-782 Added course start date to title string for instructor-led courses

### PR DESCRIPTION
**Pull request by georgebabey**
This PR will conditionally added the start date to a course when pushing course metadata to Degreed or SuccessFactors for instructor-paced courses. This will ensure the course titles are unique when there is more than one course run being published

**JIRA:** https://openedx.atlassian.net/browse/ENT-782

**Testing instructions:**
1. Create a new instructor-paced course with a start date
2. Run the transmit_course_metadata admin job
3. Expect the course to exist in the 3rd party LMS with the title as "{title} (Starts: {MM YYYY})"
4. Expect other courses with instructor-paced start dates to be updated in the 3rd party LMS after the transmit_course_metadata job has been run

**Merge checklist:**
- [x] New requirements are in the right place (`base.in` if only used in enterprise; in the correct `platform-****.in` files if they're hosted in edx-platform)
- [x] Regenerate requirements with `make upgrade && make requirements` (and make sure to fix any errors).
  **DO NOT** just add dependencies to `requirements/*.txt` files.
- [x] All reviewers approved
- [x] CI build is green
- [x] Version bumped
- [x] Changelog record added
- [x] PR author is listed in AUTHORS

**Post merge:**
- [ ] Create a tag
- [ ] Check new version is pushed to PyPi after tag-triggered build is finished.
- [ ] Delete working branch (if not needed anymore)
- [ ] edx-platform PR (be sure to include edx-platform requirements upgrades that were present in this PR)

### CODE REVIEW

File: integrated_channels/degreed/exporters/course_metadata.py [hunk 1]
```diff
@@ -185,9 +186,13 @@ def transform_cost_type(self, course_run):
...
         """
-        Return the transformed version of the course title, as well as the locale.
+        Return the transformed version of the course title. For all instructor-paced courses
+        also include the start date to distinguish multiple runs of the same course (ENT-782)
         """
-        return course_run.get('title') or ''
+        title = course_run.get('title') or ''
+        if course_run.get('pacing_type') == 'instructor_paced' and course_run.get('start'):
+            title += ' (Starts: {:%B %Y})'.format(parse_lms_api_datetime(course_run.get('start')))
+        return title
```
Comment (douglashall) [inline integrated_channels/degreed/exporters/course_metadata.py]:
```
title = course_run.get('title') or ''
course_run_start = parse_lms_api_datetime(course_run.get('start'))
if course_run_start and course_run.get('pacing_type'):
    title += ' (Starts: {:%B %Y})'.format(course_run_start)
return title
```

File: integrated_channels/sap_success_factors/exporters/course_metadata.py [hunk 2]
```diff
@@ -109,13 +110,20 @@ def transform_status(self, course_run):
             else self.STATUS_INACTIVE
         )
 
-    def transform_title(self, course_run):
```
Comment (brittneyexline) [inline integrated_channels/sap_success_factors/exporters/course_metadata.py]: is there a reason this function name changed? transform_title is consistent with what is in degreed above

File: integrated_channels/degreed/exporters/course_metadata.py (hunk 1 above)
Comment (douglashall) [reply integrated_channels/degreed/exporters/course_metadata.py]: Make sure that `parse_lms_api_datetime` can handle `None`.

File: integrated_channels/sap_success_factors/exporters/course_metadata.py [hunk 3]
```diff
@@ -109,13 +110,20 @@ def transform_status(self, course_run):
...
+    def get_title(self, course_run):
         """
-        Return the transformed version of the course title, as well as the locale.
+        Return the transformed version of the course title, as well as the locale. For all instructor-paced courses
+        also include the start date to distinguish multiple runs of the same course (ENT-782)
         """
+        title = course_run.get('title') or ''
+        if course_run.get('pacing_type') == 'instructor_paced' and course_run.get('start'):
+            title += ' (Starts: {:%B %Y})'.format(parse_lms_api_datetime(course_run.get('start')))
+        return title
```
Comment (douglashall) [inline integrated_channels/sap_success_factors/exporters/course_metadata.py]: This is the same code as what we have in the Degreed integration. Is there some way to move this to a common base class?

Comment (douglashall) [reply integrated_channels/sap_success_factors/exporters/course_metadata.py]: Also, maybe rename the shared function to something like `format_title`.

File: integrated_channels/sap_success_factors/exporters/course_metadata.py (hunk 2 above)
Comment (georgebabey) [reply integrated_channels/sap_success_factors/exporters/course_metadata.py]: Name didn't change but I pulled out a bulk of it to be reused in the `transform_content` method as well. I like Doug's suggestion of `format_title` to be a bit more descriptive than `get_title`

File: tests/test_management.py [hunk 4]
```diff
@@ -266,33 +266,33 @@ def test_transmit_course_metadata_task_success(
...
+            '"value": "edX Demonstration Course (Starts: February 2013)"}]}, {"content": [{"contentID": '
+            '"' + course_run_ids[1] + '", "contentTitle": "edX Demonstration Course (Starts: February 2013)", "launchType": 3, '
             '"launchURL": "' + settings.LMS_ROOT_URL + '/enterprise/' + uuid + '/course/'
             '' + course_run_ids[1] + '/enroll/", "mobileEnabled": false, "providerID": "SAP"}], '
             '"courseID": "' + course_run_ids[1] + '", "description": [{"locale": "English", '
             '"value": "edX Demonstration Course"}], "price": [], "providerID": "SAP", '
             '"revisionNumber": 1, "schedule": [{"active": true, "endDate": 2147483647000, '
             '"startDate": 1360040400000}], "status": "ACTIVE", "thumbnailURI": "", '
-            '"title": [{"locale": "English", "value": "edX Demonstration Course"}]}]}'
+            '"title": [{"locale": "English", "value": "edX Demonstration Course (Starts: February 2013)"}]}]}'
```
Comment (douglashall) [inline tests/test_management.py]: You could also probably make the course we are using by default for these tests self-paced so you don't have to make these changes.

File: integrated_channels/degreed/exporters/course_metadata.py (hunk 1 above)
Comment (georgebabey) [reply integrated_channels/degreed/exporters/course_metadata.py]: course_run.get('start') is guarded in the if statement so we shouldn't get `None`

File: tests/test_integrated_channels/test_sap_success_factors/test_exporters/test_course_metadata.py [hunk 5]
```diff
@@ -67,3 +67,39 @@ def test_unparsable_language_code(self):
...
+            'start': '2013-02-05T05:00:00Z',
+            'pacing_type': 'self_paced',
+            'title': 'edX Demonstration Course'
+        }
+        exporter = SapSuccessFactorsCourseExporter('fake-user', self.config)
+        assert exporter.transform_title(course_run) == \
+            [{
+                'locale': 'English',
+                'value': 'edX Demonstration Course'
+            }]
```
Comment (douglashall) [inline tests/test_integrated_channels/test_sap_success_factors/test_exporters/test_course_metadata.py]: If you have the title format code in a shared function, you probably only need one set of these tests.

File: integrated_channels/degreed/exporters/course_metadata.py (hunk 1 above)
Comment (douglashall) [reply integrated_channels/degreed/exporters/course_metadata.py]: Well, not if you do the refactor I suggested above for readability's sake and to remove the duplicate call to `course_run.get`.

**Comment by georgebabey** [comment]
@douglashall Updated - can you take another look?

File: enterprise/api_client/lms.py [hunk 6]
```diff
@@ -465,6 +465,9 @@ def parse_lms_api_datetime(datetime_string, datetime_format=LMS_API_DATETIME_FOR
         datetime_format: A datetime format string to be used for parsing
 
     """
+    if datetime_string is None:
+        return None
```
Comment (douglashall) [inline enterprise/api_client/lms.py]: On second thought, I'm not sure we want to do this. See below.

File: integrated_channels/integrated_channel/exporters/course_metadata.py [hunk 7]
```diff
@@ -104,3 +105,14 @@ def resolve_removed_courses(self, *args, **kwargs):  # pylint: disable=unused-ar
...
+    def format_title(self, course_run):
+        """
+        Return the transformed version of the course title, as well as the locale. For all instructor-paced courses
+        also include the start date to distinguish multiple runs of the same course (ENT-782)
+        """
+        title = course_run.get('title') or ''
+        course_run_start = parse_lms_api_datetime(course_run.get('start'))
+        if course_run_start and course_run.get('pacing_type') == 'instructor_paced':
+            title += ' (Starts: {:%B %Y})'.format(course_run_start)
+        return title
```
Comment (douglashall) [inline integrated_channels/integrated_channel/exporters/course_metadata.py]: One more slight tweak to get this right I think:
```
title = course_run.get('title') or ''
course_run_start = course_run.get('start')
if course_run_start and course_run.get('pacing_type') == 'instructor_paced':
    title += ' (Starts: {:%B %Y})'.format(
        parse_lms_api_datetime (course_run_start)
    )
return title
```

Comment (georgebabey) [reply integrated_channels/integrated_channel/exporters/course_metadata.py]: Perfect, that does look better, I was worried about side-effects changing parse_lms_api_datetime for the `None` case 👍

**Comment by douglashall** [comment]
@georgebabey Reminder to squash, bump version, and update changelog before merging (because I always forget).

**Pull request merged**